### PR TITLE
feat: [Qwen3-Next] Add Cute DSL GDN decode kernel and  tests

### PR DESCRIPTION
**Pull request by HongliMi**
Co-author:  @zhou9402 @liz-badada @xutizhou

## 📌 Description

This PR integrates **three versions** of the Gated Delta Rule (GDN) **Decode kernels** into FlashInfer, implemented using **CuTe DSL** for SM90 (Hopper)  and SM100 (Blackwell) GPUs. These kernels enable efficient linear attention decoding for models like Qwen3-Next.

### 🎯 Features Added

1. **Pretranspose Decode Kernel** (`gated_delta_rule_decode_pretranspose`)
   - State layout: `[B, HV, V, K]` (K-last)

### 🏗️ Architecture

All three versions follow FlashInfer's integration pattern:
- **CuTe DSL kernels** with JIT compilation and caching
- **Python API layer** with `@flashinfer_api` decorator
- **Reference implementations** for correctness validation
- **Comprehensive unit tests** with various configurations (GQA, GVA, etc.)

### 📊 Performance Highlights

**B200**（HBM3e，192GB，8TB/s）
```
GDN Decode Comparison: FlashInfer (Pretranspose) vs Triton
Config: q_heads=16, k_heads=16, v_heads=32, head_size=128, dtype=bfloat16, qk_l2norm=ON
----------------------------------------------------------------------------------------------------
 batch FlashInfer(us)   Triton(us)  FI TFLOPS  TR TFLOPS    Speedup
----------------------------------------------------------------------------------------------------
     1           4.03         5.92       0.78       0.53       1.47x
     2           4.58         6.46       1.37       0.97       1.41x
     4           5.74         7.49       2.19       1.68       1.30x
     8           8.61         9.95       2.92       2.53       1.16x
    16          14.19        16.83       3.55       2.99       1.19x
    32          24.51        31.90       4.11       3.16       1.30x
    64          49.60        58.50       4.06       3.44       1.18x
   128          92.26       113.90       4.36       3.54       1.23x
   256         175.36       225.57       4.59       3.57       1.29x
   512         340.45       442.26       4.73       3.64       1.30x
----------------------------------------------------------------------------------------------------
```

### 📁 Files Changed

**Core Implementation:**
- `flashinfer/gdn_decode.py`
  - Three CuTe DSL kernel variants
  - JIT compilation with caching
  - API functions with comprehensive validation

**Testing:**
- `tests/gdn/test_decode_delta_rule.py`
  - Unit tests for all three versions
  - Tests for various head configurations (MHA, GQA, GVA)
  - Dtype tests (float16, bfloat16)
  
- `tests/gdn/reference_delta_rule.py`
  - `decode_delta_rule()` - Single token refere

**Benchmarking:**
- `benchmarks/bench_gdn_decode.py`

### CODE REVIEW

File: flashinfer/gdn_decode.py [hunk 1]
```diff
@@ -0,0 +1,2221 @@
...
+                    offset=offset * V_PER_WARP_SMALL,
+                    mask=-1,
+                    mask_and_clamp=31,
+                )
+
+            v_new = (r_v - sum_hk) * r_beta
+            v_new = cute.arch.shuffle_sync(v_new, v_local)
+
+            sum_hq = 0.0
+            for k_iter in range(NUM_K_ITERS_SMALL, unroll=16):
```
[the remark this answers is not in the document]
Comment (yzh119) [reply flashinfer/gdn_decode.py]: It's cute-dsl syntax.

File: flashinfer/gdn_decode.py [hunk 2]
```diff
@@ -0,0 +1,2221 @@
...
+    
+    if output is None:
+        # Kernel outputs bfloat16, allocate in that dtype first
+        output = torch.zeros((B, T, HV, V), dtype=torch.bfloat16, device=q.device)
+    
+    # Convert state from [B, HV, V, K] to [B*HV, V, K] for kernel
+    h0_source = state.reshape(B * HV, V, K).contiguous()
+    
+    # Convert tensors to CuTe format
+    h0_source_tensor = from_dlpack(h0_source, assumed_align=16)
```
Comment (yzh119) [inline flashinfer/gdn_decode.py]: @HongliMi would you mind refactoring the launcher code with https://docs.nvidia.com/cutlass/latest/media/docs/pythonDSL/cute_dsl_general/compile_with_tvm_ffi.html ? It should reduce runtime overhead.

File: flashinfer/gdn_decode.py [hunk 3]
```diff
@@ -0,0 +1,2221 @@
...
+    if output is None:
+        # Kernel outputs bfloat16, allocate in that dtype first
+        output = torch.zeros((B, T, HV, V), dtype=torch.bfloat16, device=q.device)
+    
+    # State is already in K-major layout [B, HV, K, V]
+    # Use state directly as pooled view: pool_size=B, each batch is its own pool
+    h0_source = state.contiguous()
+    
+    # Convert tensors to CuTe format
+    h0_source_tensor = from_dlpack(h0_source, assumed_align=16)
```
Comment (yzh119) [inline flashinfer/gdn_decode.py]: ditto

File: flashinfer/gdn_decode.py [hunk 4]
```diff
@@ -0,0 +1,2221 @@
...
+        cache_steps = intermediate_states_buffer.shape[1]
+        intermediate_states = intermediate_states_buffer.to(torch.float32).reshape(
+            buffer_size * cache_steps * HV, V, K
+        ).contiguous()
+    else:
+        cache_steps = T
+        intermediate_states = torch.zeros(1, 1, 1, dtype=torch.float32, device=q.device)
+    
+    # Convert tensors to CuTe format
+    h0_source_tensor = from_dlpack(h0_source, assumed_align=16)
```
Comment (yzh119) [inline flashinfer/gdn_decode.py]: ditto

File: flashinfer/gdn_decode.py [hunk 5]
```diff
@@ -0,0 +1,2221 @@
...
+    # Note: kernel outputs bfloat16, we'll convert to q.dtype if needed
+    output_provided = output is not None
+    target_dtype = output.dtype if output_provided else q.dtype
+    
+    if output is None:
+        # Kernel outputs bfloat16, allocate in that dtype first
+        output = torch.zeros((B, T, HV, V), dtype=torch.bfloat16, device=q.device)
+    
+    # Convert state from [B, HV, V, K] to [B*HV, V, K] for kernel
+    h0_source = state.reshape(B * HV, V, K).contiguous()
```
Comment (yzh119) [inline flashinfer/gdn_decode.py]: Why do we need a `contiguous()` here? `reshape` itself should not affect any physical layout, it just change the shapes & strides.

File: benchmarks/bench_gdn_decode.py [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+                f"{kernel_variant:>15}"
+            )
+            
+            all_results.append(result)
+
+        print("-" * 90)
+
+
+if __name__ == "__main__":
+    main()
```
Comment (yzh119) [inline benchmarks/bench_gdn_decode.py]: nit: add an empty new line.

File: flashinfer/gdn_decode.py (hunk 2 above)
Comment (HongliMi) [reply flashinfer/gdn_decode.py]: Okay, thank you for reviewing, I have fixed it.

File: flashinfer/gdn_decode.py (hunk 5 above)
Comment (HongliMi) [reply flashinfer/gdn_decode.py]: Okay, I've removed the use of ``` contiguous()``` function.

**Review by yzh119 — APPROVED** [review]
Failed UTs are because of a bug introduced in #2366 which is fixed in https://github.com/flashinfer-ai/flashinfer/pull/2378, this PR itself should be ready to merge (and bypass the check).

**Review by vadiklyutiy — COMMENTED** [review]
@HongliMi Could you make measurement vs `fused_reccurent_gated_delta_rule` Triton kernel from vLLM?

**Comment by yzh119** [comment]
> 3.97 TB/s on H20 looks weird... The peak is 3.35TB/s

Agreed, @HongliMi can you double check?

**Comment by HongliMi** [comment]
The H100 chip has a peak memory bandwidth of 3.35TB/s, while the H20 has a peak bandwidth of 4.8TB/s.

**Comment by vadiklyutiy** [comment]
unfortunately I did not find this info on official Nvidia site.
But 3rd party says about 4TB/s as maximum...

**Comment by HongliMi** [comment]
> @HongliMi Could you make measurement vs `fused_reccurent_gated_delta_rule` Triton kernel from vLLM?

Okay, I'll add the relevant comparisons later.

**Comment by liz-badada** [comment]
> unfortunately I did not find this info on official Nvidia site. But 3rd party says about 4TB/s as maximum...

Hi, it's another version (different from 96GB, it's HBM3e, 141GB, 4.8TB/s), there is no reference, but the BW is same with H200, refer https://resources.nvidia.com/en-us-gpu-resources/hpc-datasheet-sc23

**Comment by yzh119** [comment]
I would suggest also reporting the maximum memory bandwidth of the GPU in the benchmark, you can try get_gpu_memory_bandwidth function defined in flashinfer.utils

**Comment by vadiklyutiy** [comment]
About bandwidth number. Seems in  `benchmarks/bench_gdn_decode.py` flash l2 is missed.

**Comment by vadiklyutiy** [comment]
I compared new kernels vs Triton on B200. Results are below

```
====================================================================================================
RESULTS - ALL KERNELS (Median Time in μs)
====================================================================================================

TP=4 (H=4, HV=8)
 Batch | Spec |   Triton (vllm) |    FI Decode |       FI MTP |  Triton/Decode |   Triton/MTP
----------------------------------------------------------------------------------------------------
    32 |    0 |             8.3 |         15.0 |         17.5 |          0.549 |        0.471
    32 |    1 |            12.3 |       -      |         22.8 |       -        |        0.539
    32 |    2 |            15.4 |       -      |         33.2 |       -        |        0.464
    32 |    3 |            18.6 |       -      |         41.2 |       -        |        0.452
   128 |    0 |            24.6 |         34.0 |         30.2 |          0.722 |        0.815
   128 |    1 |            37.6 |       -      |         43.4 |       -        |        0.866
   128 |    2 |            50.3 |       -      |         59.7 |       -        |        0.842
   128 |    3 |            63.3 |       -      |         76.1 |       -        |        0.833
   256 |    0 |            42.7 |         56.8 |         60.2 |          0.751 |        0.709
   256 |    1 |            65.9 |       -      |         80.6 |       -        |        0.817
   256 |    2 |            87.9 |       -      |        110.5 |       -        |        0.796
   256 |    3 |           111.8 |       -      |        141.1 |       -        |        0.792
   512 |    0 |            80.5 |        105.4 |         99.4 |          0.763 |        0.810
   512 |    1 |           125.7 |       -      |        147.9 |       -        |        0.850
   512 |    2 |           168.2 |       -      |        205.1 |       -        |        0.820
   512 |    3 |           216.0 |       -      |        268.3 |       -        |        0.805
  1024 |    0 |           157.4 |        200.0 |        177.1 |          0.787 |        0.889
  1024 |    1 |           243.4 |       -      |        278.7 |       -        |        0.873
  1024 |    2 |           326.5 |       -      |        396.6 |       -        |        0.823
  1024 |    3 |           411.4 |       -      |        520.9 |       -        |        0.790

TP=2 (H=8, HV=16)
 Batch | Spec |   Triton (vllm) |    FI Decode |       FI MTP |  Triton/Decode |   Triton/MTP
----------------------------------------------------------------------------------------------------
    32 |    0 |            14.0 |         21.1 |         21.3 |          0.667 |        0.658
    32 |    1 |            21.1 |       -      |         29.4 |       -        |        0.717
    32 |    2 |            27.6 |       -      |         41.3 |       -        |        0.669
    32 |    3 |            34.2 |       -      |         51.9 |       -        |        0.660
   128 |    0 |            42.6 |         57.1 |         59.8 |          0.747 |        0.713
   128 |    1 |            65.4 |       -      |         77.9 |       -        |        0.840
   128 |    2 |            89.3 |       -      |        110.0 |       -        |        0.812
   128 |    3 |           112.7 |       -      |        141.2 |       -        |        0.798
   256 |    0 |            81.2 |        105.1 |        100.4 |          0.772 |        0.809
   256 |    1 |           124.5 |       -      |        148.5 |       -        |        0.838
   256 |    2 |           168.3 |       -      |        205.8 |       -        |        0.818
   256 |    3 |           214.8 |       -      |        268.8 |       -        |        0.799
   512 |    0 |           157.0 |        200.5 |        178.0 |          0.783 |        0.882
   512 |    1 |           240.4 |       -      |        278.6 |       -        |        0.863
   512 |    2 |           325.0 |       -      |        396.4 |       -        |        0.820
   512 |    3 |           408.2 |       -      |        520.9 |       -        |        0.784
  1024 |    0 |           305.4 |        389.2 |        340.3 |          0.785 |        0.897
  1024 |    1 |           464.0 |       -      |        542.3 |       -        |        0.856
  1024 |    2 |           641.6 |       -      |        774.6 |       -        |        0.828
  1024 |    3 |           812.1 |       -      |       1019.4 |       -        |        0.797

TP=1 (H=16, HV=32)
 Batch | Spec |   Triton (vllm) |    FI Decode |       FI MTP |  Triton/Decode |   Triton/MTP
----------------------------------------------------------------------------------------------------
    32 |    0 |            24.5 |         33.7 |         30.6 |          0.726 |        0.802
    32 |    1 |            38.1 |       -      |         43.1 |       -        |        0.884
    32 |    2 |            51.7 |       -      |         59.6 |       -        |        0.867
    32 |    3 |            64.6 |       -      |         76.0 |       -        |        0.850
   128 |    0 |            80.6 |        105.3 |         97.2 |          0.765 |        0.829
   128 |    1 |           125.5 |       -      |        148.4 |       -        |        0.846
   128 |    2 |           169.0 |       -      |        204.4 |       -        |        0.827
   128 |    3 |           215.5 |       -      |        268.0 |       -        |        0.804
   256 |    0 |           158.6 |        199.7 |        174.8 |          0.794 |        0.907
   256 |    1 |           240.6 |       -      |        279.1 |       -        |        0.862
   256 |    2 |           324.3 |       -      |        396.5 |       -        |        0.818
   256 |    3 |           409.6 |       -      |        521.1 |       -        |        0.786
   512 |    0 |           309.7 |        388.9 |        335.2 |          0.796 |        0.924
   512 |    1 |           469.0 |       -      |        542.2 |       -        |        0.865
   512 |    2 |           636.4 |       -      |        773.6 |       -        |        0.823
   512 |    3 |           815.8 |       -      |       1028.0 |       -        |        0.794
  1024 |    0 |           606.6 |        766.5 |        657.9 |          0.791 |        0.922
  1024 |    1 |           938.8 |       -      |       1118.3 |       -        |        0.840
  1024 |    2 |          1265.5 |       -      |       1616.9 |       -        |        0.783
  1024 |    3 |          1610.8 |       -      |       2099.0 |       -        |        0.767
```

Script is here https://github.com/CentML/vllm/blob/vadim/gdn-vibe/bench_flashinfer.py

**Comment by liz-badada** [comment]
Hi @vadiklyutiy, there exists some mismatches, because we expected to compare GDN CuTe DSL kernel with this triton version (including sigmoid logics etc.): 
https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/attention/fla/fused_sigmoid_gating_recurrent.py

And it has some differences with 
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/fla/ops/fused_recurrent.py
or
https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/attention/fla/fused_recurrent.py

File: tests/gdn/conftest.py [hunk 7]
```diff
@@ -19,14 +19,19 @@
...
-        pytest.skip("GDN requires SM90a (Hopper) architecture")
+def skip_if_not_supported_arch():
+    """Skip GDN tests if not Hopper (SM90+) or Blackwell (SM100+) architecture."""
+    compute_capability = get_compute_capability(torch.device("cuda"))
+    major = compute_capability[0]
+    if major not in [9, 10, 11, 12]:
+        pytest.skip(
+            f"GDN requires Hopper (SM90+) or Blackwell (SM100+) architecture, "
+            f"but got SM{compute_capability[0]}{compute_capability[1]}"
+        )
```
[the remark this answers is not in the document]
Comment (yzh119) [reply tests/gdn/conftest.py]: We don't support CPU-only environments.

**Comment by vadiklyutiy** [comment]
Sorry for jumping in, but what’s the point of merging a kernel that’s slower than what SGLang and vLLM are already using?

**Comment by yzh119** [comment]
> Sorry for jumping in, but what’s the point of merging a kernel that’s slower than what SGLang and vLLM are already using?

Hi @vadiklyutiy thanks for the reminder, would you mind comparing with sglang's python/sglang/srt/layers/attention/fla/fused_sigmoid_gating_recurrent.py as pointed by @liz-badada .

**Comment by vadiklyutiy** [comment]
> Hi @vadiklyutiy thanks for the reminder, would you mind comparing with sglang's python/sglang/srt/layers/attention/fla/fused_sigmoid_gating_recurrent.py as pointed by @liz-badada .

I though compare vs existing implementations is mandatory part of posting PR
